File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int n;
	static int m;
	static int[] arr;
	static boolean[] isVisited;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		arr = new int[m];
		isVisited = new boolean[n + 1];

		dfs(0);
	}

	private static void dfs(int depth) {
		if (depth == m) {
			for (int num : arr) {
				System.out.print(num + " ");
			}
			System.out.println();
			return;
		}

		for (int i = 1; i <= n; i++) {
			if (!isVisited[i]) {
				isVisited[i] = true;
				arr[depth] = i;
				dfs(depth + 1);
				isVisited[i] = false;
			}
		}
	}
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys

n, m = map(int, sys.stdin.readline().rstrip().split())

arr = [0] * m
isVisited = [False] * (n + 1)


def dfs(depth):
    if depth == m:
        for num in arr:
            print(num, '', end='')
        print()
        return

    for i in range(1, n + 1):
        if isVisited[i] == False:
            isVisited[i] = True
            arr[depth] = i
            dfs(depth + 1)
            isVisited[i] = False


dfs(0)
